spinta/cli/push.py

```diff
@@ -524,33 +740,139 @@ def _add_stop_time(rows, stop):
 
 
 def _init_push_state(
-    file: pathlib.Path,
+    dburi: str,
     models: List[Model],
 ) -> Tuple[sa.engine.Engine, sa.MetaData]:
-    engine = sa.create_engine(f'sqlite:///{file}')
+    engine = sa.create_engine(dburi)
     metadata = sa.MetaData(engine)
+    inspector = sa.inspect(engine)
     for model in models:
-        table = sa.Table(
-            model.name, metadata,
-            sa.Column('id', sa.Unicode, primary_key=True),
-            sa.Column('rev', sa.Unicode),
-            sa.Column('pushed', sa.DateTime),
-        )
-        table.create(checkfirst=True)
+        need_reflect = []
+        if inspector.has_table(model.name):
+            columns = [col['name'] for col in inspector.get_columns(model.name)]
+            renamed = _rename_column(
+                engine,
+                model.name,
+                columns,
+                old_column_name='rev',
+                new_column_name='checksum'
+            )
+            need_reflect.append(renamed)
+            if not renamed:
+                need_reflect.append(_add_column(
+                    engine,
+                    model.name,
+                    columns,
+                    sa.Column('checksum', sa.Unicode)
+                ))
+            need_reflect.append(_add_column(
+                engine,
+                model.name,
+                columns,
+                sa.Column('revision', sa.Unicode)
+            ))
+            need_reflect.append(_add_column(
+                engine,
+                model.name,
+                columns,
+                sa.Column('pushed', sa.DateTime)
+            ))
+            need_reflect.append(_add_column(
+                engine,
+                model.name,
+                columns,
+                sa.Column('error', sa.Boolean)
+            ))
+            need_reflect.append(_add_column(
+                engine,
+                model.name,
+                columns,
+                sa.Column('data', sa.Text)
+            ))
+
+        if any(need_reflect):
+            metadata.reflect(only=[model.name], extend_existing=True)
```

Comment:
Methods for adding and renaming columns only update database, but they don't update metadata object, that is later used in other methods. So reflect is needed to update it, but it is called only if at least one column was actually changed in database. So if table exists and we check if column needs to be renamed or added and if it turns out that it already exists or is renamed, then reflect will not be called and it will go to table.create.

Reply:
Metadata object does not need to be updated or reflected from database, because we define it:

```python
            table = sa.Table(
                model.name, metadata,
                sa.Column('id', sa.Unicode, primary_key=True),
                sa.Column('checksum', sa.Unicode),
                sa.Column('revision', sa.Unicode),
                sa.Column('pushed', sa.DateTime),
                sa.Column('error', sa.Boolean),
                sa.Column('data', sa.Text)
            )
```

So we already have an up to date metadata, we only need to update the database schema, and for this, we either run `table.create()` or add/rename columns on already created table.

The issue here, that reflect is quite expensive operation and this adds up, when we have a large number of tables. So I would avoid doing completely unnecessary operations, especially expensive operations.